File: 10diff/076probleminc/countingsummations.py
import time 
def solution():
    print 100,len(summation(100))
    
def summation(n):
    outer = [] 
    if n == 1 or n == 2:
        inner = [] 
        inner.append(1)
        if n == 2:
            inner.append(1)
        outer.append(inner)
        return outer 
    else:
        i = n - 1
        while i > 1:
            inner = [] 
            inner.append(i)
            inner.append(n-i)
            inner.sort()
            if not inner in outer:
                outer.append(inner)
            for inners in summation(n-1):
                inner = inners 
                inner.append(1)
                inner.sort()
                if not inner in outer:
                    outer.append(inner)
            i -= 1
        return outer 
        
start = time.time()
solution()
end = time.time()
print end - start 

File: 10diff/076inc/CountingSummations.java
import java.util.ArrayList;

public class CountingSummations {
    public static void main(String[] commandLine) {
        long start = System.currentTimeMillis();
        System.out.println(count(100));
        long end = System.currentTimeMillis();
        System.out.println(end - start);
    }
    private static ArrayList<ArrayList<Integer>> count(int n) {
        ArrayList<ArrayList<Integer>> outer = new ArrayList<ArrayList<Integer>>();
        if (n == 1 || n == 2) {
            ArrayList<Integer> inner = new ArrayList<Integer>();
            inner.add(1);
            if ( n == 2 ) inner.add(1);
            outer.add(inner);
            return outer;
        }
        else {
        // fix overcounting because of unsorted inner lists
            for(int i = n-1; i > 1; i--) {
                ArrayList<Integer> inner = new ArrayList<Integer>();
                inner.add(i);
                inner.add(n-i);
                if(!outer.contains(inner)) outer.add(inner);
                for(ArrayList<Integer> inners : count(n-1)) {
                    inners.add(1);
                    if (!outer.contains(inners)) outer.add(inners);
                }
            }
            return outer;
        }
    }
}